File: modules/crewroll-transfer/android/src/main/java/com/uankit53/crewroll/transfer/media/NativeCapturePolicy.kt
package com.uankit53.crewroll.transfer.media

import java.time.Instant
import org.json.JSONObject

/** Server-owned capture boundaries; paused photos are never caught up on resume. */
internal class NativeCapturePolicy(json: JSONObject, tripId: String) {
    val version = json.getLong("version")
    val participation = json.getString("participation")
    val left = participation == "LEFT"
    val leaving = participation == "LEAVING"
    private val until = Instant.parse(json.getString("captureUntil"))
    private val excluded = json.getJSONArray("excludedCaptureWindows").let { rows ->
        (0 until rows.length()).map { index ->
            val row = rows.getJSONObject(index)
            Instant.parse(row.getString("from")) to if (row.isNull("until")) null else Instant.parse(row.getString("until"))
        }.sortedBy { it.first }
    }
    init {
        require(json.getString("tripId") == tripId && version > 0)
        require(participation in setOf("JOINED", "JOINING", "LEAVING", "LEFT"))
        require(excluded.all { it.second == null || !it.second!!.isBefore(it.first) })
    }
    fun windows(start: Instant, end: Instant): List<Pair<Instant, Instant>> {
        if (left) return emptyList()
        val bound = minOf(end, until)
        var cursor = start
        val result = mutableListOf<Pair<Instant, Instant>>()
        for ((from, to) in excluded) {
            if (to != null && to <= cursor) continue
            if (from > bound) break
            if (from > cursor) result.add(cursor to minOf(bound, from.minusMillis(1)))
            if (to == null) return result
            cursor = maxOf(cursor, to)
        }
        if (cursor <= bound) result.add(cursor to bound)
        return result
    }
}

File: modules/crewroll-transfer/ios/IdentityKeys/Sources/CrewRollNativeKeys/MediaCrypto/NativeCapturePolicy.swift
import Foundation

struct NativeCapturePolicy: Decodable {
    struct Window: Decodable { let from: String; let until: String? }
    let tripId: String
    let version: Int
    let participation: String
    var captureFrom: String? = nil
    let captureUntil: String
    let excludedCaptureWindows: [Window]
    var left: Bool { participation == "LEFT" }
    var leaving: Bool { participation == "LEAVING" }
    func windows(tripID: String, start: Date, end: Date) throws -> [(Date, Date)] {
        func date(_ string: String) throws -> Date {
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let value = formatter.date(from: string) { return value }
            formatter.formatOptions = [.withInternetDateTime]
            guard let value = formatter.date(from: string) else { throw NativeKeyError.invalidCommand }
            return value
        }
        guard tripId == tripID, version > 0, ["JOINED", "JOINING", "LEAVING", "LEFT"].contains(participation) else { throw NativeKeyError.invalidCommand }
        if left || participation == "JOINING" { return [] }
        let bound = min(end, try date(captureUntil))
        var cursor = max(start, try captureFrom.map(date) ?? start)
        var result: [(Date, Date)] = []
        for window in excludedCaptureWindows.sorted(by: { $0.from < $1.from }) {
            let from = try date(window.from)
            let to = try window.until.map(date)
            if let to, to < from { throw NativeKeyError.invalidCommand }
            if let to, to <= cursor { continue }
            if from > bound { break }
            if from > cursor { result.append((cursor, min(bound, from.addingTimeInterval(-0.001)))) }
            guard let to else { return result }
            cursor = max(cursor, to)
        }
        if cursor <= bound { result.append((cursor, bound)) }
        return result
    }
}
